lib/signed-token-request.js

```diff
@@ -0,0 +1,161 @@
+
+            return signer.sign(this.code, this.secret, requestOptions).then(() => {
+                return tokenDownloader.download(requestOptions);
+            });
+        }
+
+
+        /**
+         * Downloads content from OpenToken to a file.
+         *
+         * @param {string} token
+         * @param {string|WriteStream} stream
+         * @return {Promise.<string>}
+         */
+        downloadToFile(token, stream) {
+            var requestOptions, url;
+
+            url = requestHandler.createUrl(this.accountId, this.isPublic, token);
+            requestOptions = requestHandler.buildRequestOptions(url, "GET");
+
+            return signer.sign(this.code, this.secret, requestOptions).then(() => {
+                return tokenDownloader.downloadToFile(requestOptions, stream);
+            });
+        }
+
+
+        /**
+         * Uploads content to OpenToken.
+         *
+         * @param {string} contents
+         * @return {Promise<string>} Token for the content that was uploaded.
+         */
+        upload(contents) {
+            var requestOptions, url;
+
+            url = requestHandler.createUrl(this.accountId, this.isPublic);
+            requestOptions = requestHandler.buildRequestOptions(url, "POST");
+            requestOptions.body = contents;
+
+            return signer.sign(this.code, this.secret, requestOptions, contents).then(() => {
+                return requestAsync(requestOptions).then((response) => {
+                    var uri;
+
+                    requestHandler.checkResponse(response);
+                    uri = new Uri(response.headers.location);
+
+                    return uri.filename();
+                });
+            });
+        }
+
+
+        /**
+         * Creates an entry on OpenToken from given file.
+         *
+         * @param {string|ReadStream} stream - Path of file to load.
+         * @return {Promise<string>} Token for the content that was uploaded.
+         */
+        uploadFromFile(stream) {
+            var forkedStream, requestOptions, url;
+
+            if (typeof stream === "string") {
+                stream = fs.createReadStream(stream);
+            }
+
+            // Create a stream and when the original is read in
+            // the PassThrough stream is loaded with its data.
+            // The former is used to calculate the body hash
+            // and the latter is used for the POST request.
+            forkedStream = new StreamPassThrough();
+            stream.on("data", (chunk) => {
+                forkedStream.push(chunk);
+            });
+            stream.on("end", () => {
+                forkedStream.push(null);
+            });
+            url = requestHandler.createUrl(this.accountId, this.isPublic);
+            requestOptions = requestHandler.buildRequestOptions(url, "POST");
+
+            return signer.sign(this.code, this.secret, requestOptions, stream).then(() => {
```

Comment:
I don't see how this works without eating up memory.  What if you're trying to upload 1GB of data?